builder/penv_setup.py

```
@@ -69,6 +74,71 @@
+
+
+def _acquire_file_lock(lock_file, timeout_sec=60):
+    started = time.monotonic()
+    while True:
+        try:
+            fd = os.open(str(lock_file), os.O_CREAT | os.O_EXCL | os.O_WRONLY)
+            os.write(fd, str(os.getpid()).encode("ascii"))
+            return fd
+        except FileExistsError:
+            if time.monotonic() - started >= timeout_sec:
+                raise TimeoutError(f"Timeout waiting lock: {lock_file}")
```

Ruff (0.15.6)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)